In [ ]:
# Forage Financial Analysis: Revenue and Net Income Trends (2023–2025)

## Methodology
This notebook analyzes year-over-year (YoY) growth in revenue and net income 
for Microsoft, Tesla, and Apple using data extracted from their 10-K filings 
for fiscal years 2023–2025.

**Approach:**
1. Load the data into a pandas DataFrame from a cleaned CSV file
2. Remove empty columns left over from the Excel export
3. Convert comma-formatted text values into numeric types
4. Calculate YoY growth using `groupby('Company').pct_change()`
5. Summarize findings with aggregate statistics


In [ ]:
## Step 1: Import Libraries and Load Data

We use pandas for all data manipulation. The CSV file contains one row per 
company-year combination with seven columns covering revenue, income, assets, 
liabilities, and operating cash flow.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('forage_data.csv')
df = df.loc[:, ~df.columns.str.startswith('Unnamed')]
df.head()

,Company,Year,Total Revenue,Net Income,Total Assets,Total Liabilities,Cash Flow From Operating Activities
0,Microsoft,2023,"211,915","72,361","411,976","205,753","87,582"
1,Microsoft,2024,"245,122","88,136","512,163","243,686","118,548"
2,Microsoft,2025,"281,724","101,832","619,003","275,524","136,162"
3,Tesla,2023,"96,773","14,974","106,618","43,009","13,256"
4,Tesla,2024,"97,690","7,153","122,070","48,390","14,923"


In [ ]:
## Step 2: Data Cleaning and Type Conversion

The raw CSV values were stored as strings with comma thousands separators 
(e.g., `' 211,915 '`). We strip commas and whitespace, then convert each 
numeric column to `int64` using `pd.to_numeric`. This step is essential — 
without it, arithmetic operations on the columns would fail.

In [3]:
numeric_cols = ['Year', 'Total Revenue', 'Net Income', 
                'Total Assets', 'Total Liabilities', 
                'Cash Flow From Operating Activities']

for col in numeric_cols:
    df[col] = df[col].astype(str).str.replace(',', '', regex=False).str.strip()
    df[col] = pd.to_numeric(df[col], errors='coerce')

df.dtypes

Company                                  str
Year                                   int64
Total Revenue                          int64
Net Income                             int64
Total Assets                           int64
Total Liabilities                      int64
Cash Flow From Operating Activities    int64
dtype: object

In [ ]:
## Step 3: Year-over-Year Growth Calculation

Using `groupby('Company').pct_change()`, we compute the percentage change in 
revenue and net income for each company from one fiscal year to the next. The 
first year for each company yields `NaN` because no prior-year baseline exists.

In [4]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100
df

,Company,Year,Total Revenue,Net Income,Total Assets,Total Liabilities,Cash Flow From Operating Activities,Revenue Growth (%),Net Income Growth (%)
0,Microsoft,2023,211915,72361,411976,205753,87582,NaN,NaN
1,Microsoft,2024,245122,88136,512163,243686,118548,15.669962,21.800417
2,Microsoft,2025,281724,101832,619003,275524,136162,14.932156,15.539621
3,Tesla,2023,96773,14974,106618,43009,13256,NaN,NaN
4,Tesla,2024,97690,7153,122070,48390,14923,0.947578,-52.230533
5,Tesla,2025,94827,3855,137806,54941,14747,-2.930699,-46.106529
6,Apple,2023,383285,96995,352583,290437,110543,NaN,NaN
7,Apple,2024,391035,93736,364980,308030,118254,2.021994,-3.359967
8,Apple,2025,416161,112010,359241,285508,111482,6.425512,19.495178


In [ ]:
## Step 4: Aggregate Summary by Company

We aggregate the growth metrics by company using `.mean()` to reveal each 
firm's average performance across the two-year period.

In [5]:
df.groupby('Company')[['Revenue Growth (%)', 'Net Income Growth (%)']].mean().round(2)

,Revenue Growth (%),Net Income Growth (%)
Company,,
Apple,4.22,8.07
Microsoft,15.30,18.67
Tesla,-0.99,-49.17


## Summary of Findings

### Overview
This analysis examines year-over-year (YoY) revenue and net income growth for 
three major technology companies — Microsoft, Tesla, and Apple — across fiscal 
years 2023 through 2025.

### Microsoft — Consistent Compounder
Microsoft demonstrated the most reliable growth. Revenue increased **15.67%** 
in 2024 and **14.93%** in 2025, while net income grew **21.80%** and **15.54%** 
respectively. Consistent top-line and bottom-line growth suggests strong 
operational execution.

### Tesla — Deteriorating Profitability
Tesla showed the most concerning trend. Revenue was essentially flat (**+0.95%**) 
in 2024 and turned negative in 2025 (**-2.93%**). More alarmingly, net income 
plummeted **-52.23%** in 2024 and a further **-46.11%** in 2025.

### Apple — V-Shaped Recovery
Apple's revenue grew modestly (**+2.02%** in 2024, **+6.43%** in 2025). Net 
income dipped slightly in 2024 (**-3.36%**) but rebounded strongly in 2025 
(**+19.50%**) — the highest single-year net income growth across all three 
companies.

### Key Insights
- **Best overall performer:** Microsoft — balanced, consistent growth.
- **Biggest concern:** Tesla — revenue stagnation + two-year net income collapse.
- **Most improved:** Apple — 2025 net income rebound.